packages/combobox/src/useCombobox.ts

```diff
@@ -0,0 +1,777 @@
+    [
+      getDownshiftTriggerProps,
+      getDownshiftInputProps,
+      getFieldInputProps,
+      triggerRef,
+      disabled,
+      _selectionValue,
+      _isExpanded,
+      _activeIndex,
+      closeListbox,
+      openListbox,
+      setActiveIndex,
+      matchValue,
+      values,
+      labels,
+      triggerContainsInput,
+      isAutocomplete,
+      isEditable,
+      inputRef
+    ]
+  );
+
+  const getLabelProps = useCallback<IUseComboboxReturnValue['getLabelProps']>(
+    ({ onClick, ...other } = {}) => {
+      const { htmlFor, ...labelProps } = getFieldLabelProps(other);
+      const handleClick = () => !isEditable && triggerRef.current?.focus();
+
+      return {
+        ...labelProps,
+        onClick: composeEventHandlers(onClick, handleClick),
+        htmlFor: isEditable ? htmlFor : undefined
+      };
+    },
+    [getFieldLabelProps, isEditable, triggerRef]
+  );
+
+  const getInputProps = useCallback<IUseComboboxReturnValue['getInputProps']>(
+    ({
+      role = isEditable ? 'combobox' : null,
+      'aria-labelledby': ariaLabeledBy = null,
+      onClick,
+      onFocus,
+      ...other
+    } = {}) => {
+      const inputProps = {
+        'data-garden-container-id': 'containers.combobox.input',
+        'data-garden-container-version': PACKAGE_VERSION,
+        ref: inputRef,
+        role: role === null ? undefined : role,
+        onClick,
+        onFocus
+      };
+
+      if (isEditable) {
+        const handleClick = (event: MouseEvent) =>
+          event.target instanceof Element &&
+          triggerRef.current?.contains(event.target) &&
+          event.stopPropagation();
+
+        return getDownshiftInputProps({
+          ...inputProps,
+          disabled,
+          role,
+          'aria-labelledby': ariaLabeledBy,
+          'aria-autocomplete': isAutocomplete ? 'list' : undefined,
+          onClick: composeEventHandlers(onClick, handleClick),
+          ...getFieldInputProps(),
+          ...other
+        } as IDownshiftInputProps);
+      }
+
+      const downshiftInputProps = getDownshiftInputProps({
+        ...inputProps,
+        disabled: true,
+        'aria-autocomplete': undefined,
+        'aria-activedescendant': undefined,
+        'aria-controls': undefined,
+        'aria-expanded': undefined,
+        'aria-hidden': true,
+        'aria-labelledby': undefined
+      });
+
+      const handleFocus = () => {
+        if (!isEditable) {
+          // Prevent Downshift from grabbing non-editable combobox focus.
+          triggerRef.current?.focus();
+        }
+      };
+
+      // Select-only combobox input
+      return {
+        ...downshiftInputProps,
+        disabled,
+        readOnly: true,
+        tabIndex: -1,
+        onFocus: composeEventHandlers(onFocus, handleFocus),
+        ...other
+      };
+    },
+    [
+      getDownshiftInputProps,
+      getFieldInputProps,
+      inputRef,
+      triggerRef,
+      disabled,
+      isAutocomplete,
+      isEditable
```

Comment:
Is there much guarantee that these dependencies won't cause frequent re-memoizations from React?

Once these lists get long the perf benefit gets questionable, but at least downshift [wraps their getters](https://github.com/downshift-js/downshift/blob/master/src/hooks/useCombobox/index.js#L387) in `useCallback`, which is nice.